src/game/logic/object/update/laserupdate.cpp

```diff
@@ -0,0 +1,365 @@
+            if (data->m_muzzleParticleSystem.Is_Not_Empty()) {
+                ParticleSystemTemplate *tmplate = g_theParticleSystemManager->Find_Template(data->m_muzzleParticleSystem);
+
+                if (tmplate) {
+                    ParticleSystem *system = g_theParticleSystemManager->Create_Particle_System(tmplate, true);
+
+                    if (system) {
+                        m_muzzleParticleSystemID = system->Get_System_ID();
+                    }
+                }
+            }
+
+            if (data->m_targetParticleSystem.Is_Not_Empty()) {
+                ParticleSystemTemplate *tmplate = g_theParticleSystemManager->Find_Template(data->m_targetParticleSystem);
+
+                if (tmplate) {
+                    ParticleSystem *system = g_theParticleSystemManager->Create_Particle_System(tmplate, true);
+
+                    if (system) {
+                        m_targetParticleSystemID = system->Get_System_ID();
+                    }
+                }
+            }
+        }
+    }
+
+    if (m_muzzleParticleSystemID) {
+        ParticleSystem *system = g_theParticleSystemManager->Find_Particle_System(m_muzzleParticleSystemID);
+
+        if (system) {
+            system->Set_Position(m_startPos);
+        }
+    }
+
+    if (m_targetParticleSystemID) {
+        ParticleSystem *system = g_theParticleSystemManager->Find_Particle_System(m_targetParticleSystemID);
+
+        if (system) {
+            system->Set_Position(m_endPos);
+        }
+    }
+
+    Coord3D pos;
+
+    if (source_obj) {
+        pos = *source_obj->Get_Position();
+    } else {
+        pos.Set(start_pos);
+        pos.Add(end_pos);
+        pos.Scale(0.5f);
+    }
+
+    Drawable *drawable = Get_Drawable();
+
+    if (drawable) {
+        drawable->Set_Position(&pos);
+    }
+
+    m_dirty = true;
+}
+
+void LaserUpdate::Update_End_Pos()
+{
+    const LaserUpdateModuleData *data = Get_Module_Data();
+    Coord3D oldpos = m_endPos;
+
+    if (m_victimDrawableID) {
+        Drawable *drawable = g_theGameClient->Find_Drawable_By_ID(m_victimDrawableID);
+        bool dead;
+
+        if (drawable && drawable->Get_Object()) {
+            dead = drawable->Get_Object()->Is_Effectively_Dead();
+        } else {
+            dead = false;
+        }
+
+        if (drawable && !dead) {
+            m_endPos = *drawable->Get_Position();
+        } else {
+            if (data->m_punchThroughScalar) {
+                Vector3 pos;
+                pos.Set(m_endPos.x, m_endPos.y, m_endPos.z);
+                pos = pos - Vector3(m_startPos.x, m_startPos.y, m_startPos.z);
+                pos *= data->m_punchThroughScalar;
+                pos = pos + Vector3(m_startPos.x, m_startPos.y, m_startPos.z);
+                m_endPos.x = pos.X;
+                m_endPos.y = pos.Y;
+                m_endPos.z = pos.Z;
+            }
+
+            m_victimDrawableID = DRAWABLE_UNK;
+        }
+
+        if (!(m_startPos == oldpos)) {

```

Comment:
Is there a reason this is written as `!(m_startPos == oldpos)` as opposed to `(m_startPos != oldpos)` ? I suspect **operator!=** was not defined.

Reply:
The original code did it this way (It called operator== and used ! on the result)